### ISSUE TITLE
[explicit-function-return-type] False negative - Methods of nested objects

### ISSUE THREAD

**Issue by beaumontjonathan** [typescript-eslint/typescript-eslint#1119]
**Description**

`explicit-function-return-type` does not seem to respect nested object types, whether the nested object is an argument or just regular argument.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-function-return-type": "warn"
  }
}
```

```TS
declare function functionWithNestedObjectArg(arg: { nested: { method: () => number } });
functionWithNestedObjectArg({
  nested: {
    method() {
      return 1;
    },
  },
});

const foo: {
  bar: {
    baz: () => string;
  };
} = {
  bar: {
    baz: () => '',
  },
};
```

**Expected Result**

No warning

```sh
dir $ yarn eslint --ext ts,js src/reproduce.ts 
yarn run v1.19.0
$ /Users/jonathan/dir/node_modules/.bin/eslint --ext ts,js src/reproduce.ts
✨  Done in 1.32s.
```

**Actual Result**

```sh
dir $ yarn eslint --ext ts,js src/reproduce.ts 
yarn run v1.19.0
$ /Users/jonathan/dir/node_modules/.bin/eslint --ext ts,js src/reproduce.ts

/Users/jonathan/dir/src/reproduce.ts
   4:5   warning  Missing return type on function  @typescript-eslint/explicit-function-return-type
  16:10  warning  Missing return type on function  @typescript-eslint/explicit-function-return-type

✖ 2 problems (0 errors, 2 warnings)

✨  Done in 1.32s.
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.3` |
| `@typescript-eslint/parser`        | `2.3.3` |
| `TypeScript`                       | `3.6.4` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.7.0` |

**Comment by bradzacher** [comment]
IIRC I explicitly did not go into handling these sorts of cases, because things can get really complicated with the types, and the explicit return type contract can get pretty fuzzy once you go beyond the first level.

```ts
type CB = () => void;
type MyObj = {
  prop: CB;
};
type NestedMyObj1 = {
  prop: MyObj
};
type NestedMyObj2 = {
  prop: MyObj
};
// etc etc

function foo(arg: MyOtherObjN): void {}

foo({
  prop: {
    prop: {
      // ...
      prop: () => {}
      // should this be "allowed" by the rule? its return type is checked, sure.
      // but it's pretty much impossible to exactly interrogate it, or infer
      // what it is supposed to be.
    }
  }
});
```

Additionally, objects can be really easily weakly typed records `Record<string, unknown>`, and as this rule operates without type information (I.e. has to make its assumptions purely based on the AST), it gets harder to assume that the function is explicitly typed:

```ts
// these cases are easy to assume the function is typed in these cases
foo(() => {});
const bar: MyType = () => {};
const baz = () => {} as MyType;
class Foo {
  bar: MyType = () => {}
}
const x: MyObjType = {
  prop: () => {},
};

// this is hard to assume the function is typed
const x: MyNestedObjType = {
  prop: {
    prop: {
      prop: {
        prop: {
          prop: () => {},
        }
      }
    }
  }
}
```

Should this deeply-nested case be "allowed" by the rule?
I mean, sure, its return type _might_ be explicitly checked by TS itself.
However in a lot of cases it can be pretty hard to exactly interrogate its type within an IDE, or more importantly for a reviewer outside of an IDE.

Which is kind of the point of the rule - creating explicit contracts that are easy to see, and follow - to ensure type correctness.
